File: src/uniproxy/clash/conf.py
from __future__ import annotations

from typing import Literal, Mapping, Sequence, TypeAlias

from attrs import define

from .base import (
    AbstractClash,
    BaseProtocol,
    BaseProxyGroup,
    BaseRule,
    BaseProxyProvider,
)
from .rules import RuleProvider

Hosts: TypeAlias = Mapping[str, str]
Proxies: TypeAlias = Sequence[BaseProtocol]
ProxyProviders: TypeAlias = Sequence[BaseProxyProvider]
ProxyGroups: TypeAlias = Sequence[BaseProxyGroup]
RuleProviders: TypeAlias = Sequence[RuleProvider]
Rules: TypeAlias = Sequence[BaseRule]

Mode = Literal["rule", "global", "direct"]
LogLevelType = Literal["silent", "info", "warning", "error", "debug"]


@define
class ClashConfig(AbstractClash):
    mode: Mode
    log_level: LogLevelType
    ipv6: bool

    port: int
    socks_port: int
    redir_port: int
    mixed_port: int

    allow_lan: bool
    bind_address: str
    external_controller: str

    # dns: Dns
    # tun: Tun
    # hosts: Hosts
    proxies: Proxies
    proxy_providers: ProxyProviders
    proxy_groups: ProxyGroups
    rule_providers: RuleProviders
    rules: Rules

File: uniproxy/src/clash/conf.scala
package uniproxy
package clash
package conf

import com.comcast.ip4s.{Host, Port, SocketAddress}
import com.comcast.ip4s.IpAddress

import uniproxy.typing.{GroupType, Network, ProtocolType}
import uniproxy.clash.abc.{
  AbstractClash,
  ProtocolLike,
  ProxyGroupLike,
  ProxyProviderLike,
  RuleProviderLike,
}
import uniproxy.clash.rules.Rule

// Type aliases using literal types
type Mode = "rule" | "global" | "direct"
type LogLevelType = "silent" | "info" | "warning" | "error" | "debug"

// Collection type aliases
type Hosts = Map[String, String]
type Proxies = Seq[ProtocolLike]
type ProxyProviders = Seq[ProxyProviderLike]
type ProxyGroups = Seq[ProxyGroupLike]
type RuleProviders = Seq[RuleProviderLike]
type Rules = Seq[Rule]

// Main configuration case class
case class ClashConfig(
  mode: Mode,
  logLevel: LogLevelType,
  ipv6: Boolean,

  // Network settings
  port: Port,
  socksPort: Int,
  redirPort: Int,
  mixedPort: Int,
  allowLan: Boolean,
  bindAddress: String,

  // External controller settings
  externalController: SocketAddress[IpAddress],
  externalUI: String,
  externalUiUrl: String,
  secret: String,

  // proxy
  proxies: Proxies = Seq.empty,
  proxyProviders: ProxyProviders = Seq.empty,
  proxyGroups: ProxyGroups = Seq.empty,

  // rules
  ruleProviders: RuleProviders = Seq.empty,
  rules: Rules = Seq.empty,
) extends AbstractClash

// // Helper constants if needed
// object ClashConfig {
//   val Modes = Map(
//     "Rule" -> "rule",
//     "Global" -> "global",
//     "Direct" -> "direct"
//   )

//   val LogLevels = Map(
//     "Silent" -> "silent",
//     "Info" -> "info",
//     "Warning" -> "warning",
//     "Error" -> "error",
//     "Debug" -> "debug"
//   )
// }
